scripts/staging/fm-binclass.dml

```diff
@@ -0,0 +1,179 @@
+   * fm::init -> adam::init -> fm::forward -> sigmoid::forward -> log_loss::forward \
+   *          adam::update <- fm::backward <- sigmoid::backward <- log_loss::backward <-
+   */
+
+    n = nrow(X);
+    d = ncol(X);
+    k = 2; # factorization dimensionality, only(=2) possible for now.
+ 
+    # 1.initialize fm core
+    [w0, W, V] = fm::init(n, d, k);
+    
+    # 2.initialize adam optimizer
+    ## Default values for some parameters
+    lr      = 0.001; 
+    beta1   = 0.9;       # [0, 1)
+    beta2   = 0.999;     # [0, 1)
+    epsilon = 0.00000001; 
+    t       = 0;
+    
+    # [mX, vX] = adam::init(X); # to optimize input.
+    [mw0, vw0] = adam::init(w0);
+    [mW, vW]   = adam::init(W);
+    [mV, vV]   = adam::init(V);
+    
+    # Regularization
+    lambda = 5e-04
+
+    # Optimize
+    print("Starting optimization")
+    batch_size = 10
+    iters = ceil(1000 / batch_size)
+    epochs = 100; N = 1000
+    for (e in 1:epochs) {
+      for (i in 1:iters) {
+        # Get the next batch
+        beg = ((i-1) * batch_size) %% N + 1
+        end = min(N, beg + batch_size - 1)
+        X_batch = X[beg:end,]
+        y_batch = y[beg:end,]
+    
+        # 3.Send inputs through fm::forward
+        y_res = fm::forward(X_batch, w0, W, V);
+    
+        # 4.Send the above result through sigmoid::forward
+        sfy = sigmoid::forward(y_res);
+    
+        # 5.Send the above result through log_loss::forward
+        loss = log_loss::forward(sfy, y_batch);
+
+        # Compute loss & accuracy for training & validation data every 100 iterations.
+        if (i %% 100 == 0) {
+          # Compute training loss & accuracy
+          loss_data = log_loss::forward(sfy, y_batch);
+          loss_reg_X  = l2_reg::forward(X, lambda);
+          loss_reg_w0 = l2_reg::forward(w0, lambda);
+          loss_reg_W  = l2_reg::forward(W, lambda);
+          loss_reg_V  = l2_reg::forward(V, lambda);
+          
+          accuracy = mean((loss_data<0.5) == (y_batch<0.5));
+          loss = loss_data + loss_reg_X + loss_reg_w0 + loss_reg_W + loss_reg_V;
+          
+          # Compute validation loss & accuracy
+          probs_val = predict(X_val, w0, W, V)
+          loss_val = log_loss::forward(probs_val, y_val)
+          accuracy_val = mean((probs_val<0.5) == (y_val<0.5))
+
+          # Output results
+          print("Epoch: " + e + ", Iter: " + i + ", Train Loss: " + loss + ", Train Accuracy: "
+                + accuracy + ", Val Loss: " + loss_val + ", Val Accuracy: " + accuracy_val)
+        }
+        
+        # 6.Send the result of sigmoid::forward and the correct labels y to log_loss::backward
+        dsfy = log_loss::backward(sfy, y_batch);
+    
+        # 7.Send the above result through sigmoid::backward
+        dy = sigmoid::backward(dsfy, y_res);
+    
+        # 8.Send the above result through fm::backward
+        [dw0, dW, dV] = fm::backward(dy, X_batch, w0, W, V);
+        
+        # 9.Call adam::update for all parameters
```

Comment:
Please update the timestep `t` variable here before passing it into Adam: `t = e*i - 1`.